### ISSUE TITLE
Make large mutations faster

### ISSUE THREAD

**Issue by jhalborg** [prisma/prisma#173]
Hello Prisma Team, 

Thank you for a great product, so excited about v2! 👏

We're using Prisma 1 in production on a fairly large codebase with a couple of thousand monthly users. 

One of our biggest issues with Prisma 1 is performance. There are several factors for this, many of which will be improved upon with Prisma 2, but there's one aspect I'm unsure if Prisma 2 has a solution for (yet). 

When looking at our Apollo Engine data for the graph, 90% of our slowest p95 operations are mutations, and even fairly simple mutations are pretty slow. For medium to large mutations, we're no longer measuring in ms, but in seconds. We've also set up Datadog tracing for our resolvers, and have found that the calls to our Prisma server accounts for about 60-80% of the resolve time for mutations on average. 

I was wondering if Prisma 2 has a way to improve performance of mutations, other than the faster Rust implementation? The problem is especially grim for larger mutations that affect multiple models. 

One solution is to break down the updates into separate mutations and perform them in parallel instead of as one big query, but this messes with atomicity. Perhaps there could be a way to pass in an array of mutations to be performed only if they all succeed?

Thanks 😄

**Comment by jhalborg** [comment]
@pantharshit00 - Does this issue in fact belong in `prisma/photonjs` instead?

**Comment by pantharshit00** [comment]
It belongs here. This is related to the query engine.

**Comment by mavilein** [comment]
@jhalborg : Can you show exampales for each category of mutation you are talking about? So one for simple, one for medium and one for large? This would help a lot 🙏

**Comment by jhalborg** [comment]
I'll try, @mavilein , 

We are bindings users, but the underling query engine should be the same as Prisma Client I believe?

Note: I have made slight modifications to not expose some info I don't want exposed, but it should get the point across. I hope it makes sense. 

**Simple**
```
await ctx.db.mutation.createPushNotificationToken(
      { data: { version: 1, token, user: { connect: { id: user.id } } } },
      '{ id }',
    );
```

**Medium**
```
await ctx.db.mutation.updateUser(
    {
      where: { id: userId },
      data: {
        version: user.version + 1,
        details: {
          delete: isDefined(user.details),
        },
        emailVerification: {
          delete: isDefined(user.emailVerification),
        },
        application: {
          delete: isDefined(user.application),
        },
        passwordReset: {
          delete: isDefined(user.passwordReset),
        },
        proposedUserFor: user.proposedUserFor
          ? {
              disconnect: user.proposedUserFor.map(({ id }) => ({ id })),
            }
          : undefined,
        pushNotificationTokens: user.pushNotificationTokens
          ? {
              delete: user.pushNotificationTokens.map(({ id }) => ({ id })),
            }
          : undefined,
        refreshTokens: user.refreshTokens
          ? {
              delete: user.refreshTokens.map(({ id }) => ({ id })),
            }
          : undefined,
        selectedUserFor: user.selectedUserFor
          ? {
              disconnect: user.selectedUserFor.map(({ id }) => ({ id })),
            }
          : undefined,
        userCredentials: { delete: isDefined(user.userCredentials) },
        userRights: user.userRights ? { delete: user.userRights.map(({ id }) => ({ id })) } : undefined,
        userComments: user.userComments ? { delete: user.userComments.map(({ id }) => ({ id })) } : undefined,
        userCommentsAuthored: user.userCommentsAuthored
          ? { delete: user.userCommentsAuthored.map(({ id }) => ({ id })) }
          : undefined,
        campaignDraftComments: user.campaignDraftComments
          ? { delete: user.campaignDraftComments.map(({ id }) => ({ id })) }
          : undefined,
        successAgentFor: user.successAgentFor
          ? { disconnect: user.successAgentFor.map(({ id }) => ({ id })) }
          : undefined,
        agentForCompanies: user.agentForCompanies
          ? { disconnect: user.agentForCompanies.map(({ id }) => ({ id })) }
          : undefined,
        reservations: user.reservations
          ? { delete: user.reservations.map(({ id }) => ({ id })) }
          : undefined,
        orders: user.orders ? { disconnect: user.orders.map(({ id }) => ({ id })) } : undefined,
      },
    },
    info,
  );
```

**Heavy**
This is an unfortunate mutation, but not sure how we could improve it. It copies the data from one rather large model (campaign) into a new entry, so the first works as a "template" to build a new one off of. 

```
await ctx.db.mutation.createCampaign(
    {
      data: {
        version: 1,
        status: CampaignPublicationState.underConstruction,
        openForSubmissions: false,
        audienceType: campaign.audienceType,
        budget: campaign.budget,
        company: { connect: { id: company.id } },
        createdBy: { connect: { id: ctx.accessContext.getUserId() } },
        emailsToNotify: { set: campaign.emailsToNotify },
        name: campaign.name,
        draftMeta: {
          create: {
            version: 1,
            status: CampaignDraftState.DRAFT,
            draft: {
              create: {
                version: 1,
                createdBy: { connect: { id: ctx.accessContext.getUserId() } },
                showInSubmissionAppFeed: draft.showInSubmissionAppFeed || true,
                companyRepulsions: draft.companyRepulsions
                  ? {
                      create: draft.companyRepulsions.map<CompanyRepulsionCreateInput>((item) => ({
                        version: 1,
                        company: { connect: { id: item.company.id } },
                        repulsion: item.repulsion,
                      })),
                    }
                  : undefined,
                audienceType: draft.audienceType,
                defaultAccountTags: { set: draft.defaultAccountTags },
                defaultCaption: draft.defaultCaption,
                defaultTags: { set: draft.defaultTags },
                description: draft.description,
                endDate: draft.endDate,
                genders: { set: draft.genders },
                inspirationImages: draft.inspirationImages
                  ? { connect: draft.inspirationImages.map(({ id }) => ({ id })) }
                  : undefined,
                interestRepulsions: draft.interestRepulsions
                  ? {
                      create: draft.interestRepulsions.map<InterestRepulsionCreateInput>((item) => ({
                        version: 1,
                        interest: { connect: { id: item.interest.id } },
                        repulsion: item.repulsion,
                      })),
                    }
                  : undefined,
                interests: draft.interests ? { connect: draft.interests.map(({ id }) => ({ id })) } : undefined,
                isGlobal: draft.isGlobal,
                location: draft.location
                  ? {
                      create: {
                        city: draft.location.city,
                        country: draft.location.country,
                        latitude: draft.location.latitude,
                        longitude: draft.location.longitude,
                        placeId: draft.location.placeId,
                        street: draft.location.street,
                        streetNumber: draft.location.streetNumber,
                      },
                    }
                  : undefined,
                maximumAge: draft.maximumAge,
                minimumAge: draft.minimumAge,
                objectives: draft.objectives ? { connect: draft.objectives.map(({ id }) => ({ id })) } : undefined,
                endDate: draft.endDate,
                startDate: draft.startDate,
                postMedia: { set: draft.postMedia },
                radiusInMeters: draft.radiusInMeters,
                selectedUsers: {
                  connect: [
                    ...(draft.selectedUsers || []).map((selectedUser) => selectedUser.id),
                    ctx.accessContext.getUserRoles().includes(UserRole.employee)
                      ? ctx.accessContext.getUserId()
                      : undefined,
                  ]
                    .filter(isDefined)
                    .filter((v, i, a) => a.indexOf(v) === i)
                    .map((id) => ({ id })),
                },
                startDate: draft.startDate,
                subtitle: draft.subtitle,
                title: draft.title,
                visibleBeforeStartDate: draft.visibleBeforeStartDate,
                watermarks: draft.watermarks ? { connect: draft.watermarks.map(({ id }) => ({ id })) } : undefined,
                weight: draft.weight,
                rewards: draft.rewards
                  ? {
                      create: draft.rewards.map<CampaignDraftSubmissionRewardCreateWithoutCampaignDraftInput>(
                        (reward) => ({
                          version: 1,
                          postMediaType: reward.postMediaType,
                          voucherValue: reward.voucherValue,
                          brandcoins: reward.brandcoins,
                          note: reward.note,
                        }),
                      ),
                    }
                  : undefined,
              },
            },
          },
        },
        submissionPrices: campaign.submissionPrices
          ? {
              create: campaign.submissionPrices.map((submissionPrice) => ({
                version: 1,
                price: submissionPrice.price,
                postMediaType: submissionPrice.postMediaType,
              })),
            }
          : undefined,
      },
    },
    info,
  );
```

**Comment by mavilein** [comment]
@jhalborg : Have you maybe tried this mutation with one of the recent versions of Prisma 2?

**Comment by jhalborg** [comment]
@mavilein - I haven't yet, but I've shared our database schema with the Prisma team via schemas@prisma.io so you can easily set up some tests that you think makes sense based on that datamodel if you think it can help you guys with this issue? 💪

**Comment by janpio** [comment]
(Internal note: Private repo, prisma1-old, "Jesper Madsen")

**Comment by janpio** [comment]
This needs to be treated as a bug in our implementation, the underlying schema and the GraphQL API mutations above could give some indication what to test for and look at here. As quite some time has passed, it might also be that all of this is outdated and not actually a problem any more.

**Comment by igalklebanov** [comment]
Hey 👋 

This one's dusty and irrelevant since we moved away from Rust in v7. If anyone's experiencing this issues with Prisma 7, please submit a new issue.
